lib/ab_test.rb

```diff
@@ -0,0 +1,107 @@
+# frozen_string_literal: true
+
+class AbTest
+  attr_reader :buckets, :experiment_name, :default_bucket, :should_log
+
+  MAX_SHA = (16 ** 64) - 1
+
+  # @param [Proc<String>,RegExp,string,Boolean,nil] should_log Controls whether bucket data for this
+  #                                                            A/B test is logged with specific
+  #                                                            events.
+  # @yieldparam [ActionDispatch::Request] request
+  # @yieldparam [String,nil] service_provider Issuer string for the service provider associated with
+  #                                           the current session.
+  # @yieldparam [User] user
+  # @yieldparam [Hash] user_session
+  def initialize(
+    experiment_name:,
+    buckets: {},
+    should_log: nil,
+    default_bucket: :default,
+    &discriminator
+  )
+    @buckets = buckets
+    @discriminator = discriminator
+    @experiment_name = experiment_name
+    @default_bucket = default_bucket
+    @should_log = should_log
+    raise 'invalid bucket data structure' unless valid_bucket_data_structure?
+    ensure_numeric_percentages
+    raise 'bucket percentages exceed 100' unless within_100_percent?
+  end
+
+  # @param [ActionDispatch::Request] request
+  # @param [String,nil] service_provider Issuer string for the service provider associated with
+  #                                      the current session.
+  # @params [Hash] session
+  # @param [User] user
+  # @param [Hash] user_session
+  def bucket(request:, service_provider:, session:, user:, user_session:)
+    return nil if no_percentages?
+
+    discriminator = resolve_discriminator(
+      request:, service_provider:, session:, user:,
+      user_session:
+    )
+    return nil if discriminator.blank?
+
+    user_value = percent(discriminator)
+
+    min = 0
+    buckets.keys.each do |key|
+      max = min + buckets[key]
+      return key if user_value > min && user_value <= max
+      min = max
+    end
+
+    @default_bucket
+  end
+
+  def include_in_analytics_event?(event_name)
+    if should_log.is_a?(Regexp)
+      should_log.match?(event_name)
+    elsif !should_log.nil?
+      raise 'Unexpected value used for should_log'
+    else
+      true
+    end
+  end
+
+  private
+
+  def resolve_discriminator(user:, **)
+    if @discriminator
+      @discriminator.call(user:, **)
+    elsif !user.is_a?(AnonymousUser)
+      user&.uuid
```

Comment:
We previously used the user's uuid + the A/B test name - this keeps a user from always landing in the same percentage. e.g., if someone's uuid always resolved to 1, they'd always end up in the first bucket for every A/B test.
```suggestion
      user&.uuid + experiment_name
```

Reply:
Currently this gets handled down in `percent`: https://github.com/18F/identity-idp/blob/3a4090abf22e93807014c98fefa8336980b5d6d6/lib/ab_test.rb#L85